python/sglang/srt/layers/sampler.py

```diff
@@ -112,7 +115,11 @@ def forward(
 
             # Post process logits
             logits.div_(sampling_info.temperatures)
-            logits[:] = torch.softmax(logits, dim=-1)
+            # For ascend backend, softmax is not needed before sampling
+            if not get_global_server_args().sampling_backend == "ascend" or (
+                return_logprob and not SGLANG_RETURN_ORIGINAL_LOGPROB
+            ):
+                logits[:] = torch.softmax(logits, dim=-1)
```

Comment:
Because softmax of the raw logits is included in the fused op, the initial softmax can be skipped if `return_logprob` is false. Condition is copied from here https://github.com/sgl-project/sglang/blob/dc4f54182374e9a52cefdd0b96eb468b40f04529/python/sglang/srt/layers/sampler.py#L158-L169